### PR TITLE
Practice 1: Loan interest calculator

### PR DESCRIPTION
**Pull request by Tang2812**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,6 @@
 #ignore all files in .idea
```
Comment (ghost) [inline .gitignore:1]: Please help ignore it.
image

File: practice/dist/index.html [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            <th class="table__header-footer-content" id="total-payment">0</th>
+          </tr>
+        </tfoot>
+
+      </table>
+      <button id="export">Export to excel</button>
+
+    </div>
+  </div>
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/xlsx/0.17.4/xlsx.full.min.js"></script>
```
Comment (ghost) [inline practice/dist/index.html]: you can remove it and install by npm

File: practice/dist/index.html [hunk 3]
```diff
@@ -0,0 +1,163 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+  <meta charset="UTF-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <link rel="stylesheet" href="/index.3092e1e0.css">
+  <link rel="stylesheet" href="https://code.jquery.com/ui/1.12.1/themes/base/jquery-ui.css">
```
Comment (ghost) [inline practice/dist/index.html]: remove jquery

File: practice/dist/index.html [hunk 4]
```diff
@@ -0,0 +1,163 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+  <meta charset="UTF-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <link rel="stylesheet" href="/index.3092e1e0.css">
```
Comment (ghost) [inline practice/dist/index.html]: please check

File: practice/dist/index.html [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  </div>
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/xlsx/0.17.4/xlsx.full.min.js"></script>
+  <script src="/index.3d4006dd.js"></script>
+  <script src="/index.1e62695a.js"></script>
+  <script src="/index.812cfa71.js"></script>
+  <script src="/index.2c921054.js"></script>
+  <script src="/index.13e98d19.js"></script>
+  <script src="/index.e2e7fb24.js"></script>
+  <script src="/index.aa69868b.js"></script>
+<script src="/practice.59712999.js"></script></body>
```
Comment (ghost) [inline practice/dist/index.html]: Please check

File: practice/src/index.html [hunk 6]
```diff
@@ -0,0 +1,168 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+  <meta charset="UTF-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <link rel="stylesheet" href="./css/index.css">
+  <link rel="stylesheet" href="https://code.jquery.com/ui/1.12.1/themes/base/jquery-ui.css">
```
Comment (ghost) [inline practice/src/index.html]: remove jquery?

File: practice/src/js/controllers/form_input_controller.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    } else {
+      inputErrors.errorMessages.push('');
+    };
+
+    // validate Disbursement date
+    const date = PageLoadController.getDayToDay();
+    const [day, month, year] = date.split('/').map(Number);
+    const [inputDay, inputMonth, inputYear] = valueOfDisbursementDate.split('/').map(Number);
+
+    if (inputYear < year) {
```
Comment (ghost) [inline practice/src/js/controllers/form_input_controller.js]: refactor nested if else

File: practice/src/js/controllers/form_input_controller.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  },
+
+  // function check leap year
+  isLeapYear: function (year) {
+    return (year % 4 === 0 && year % 100 !== 0) || (year % 400 === 0);
+  },
+
+  // function calculate number of date in a month
+  getDaysInMonth: function (month, year) {
+    return [31, (this.isLeapYear(year) ? 29 : 28), 31, 30, 31, 30, 31, 31, 30, 31, 30, 31][month - 1];
```
Comment (ghost) [inline practice/src/js/controllers/form_input_controller.js]: Should create const value for numbers

File: practice/src/js/controllers/form_input_controller.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+  calculateRepaymentDate: function (date) {
+    const [day, month, year] = date.split('/').map(Number);
+    const daysOfMonth = this.getDaysInMonth(month, year);
+    let newMonth = month;
+    let newDay = day;
+    let newYear = year;
+    let newDate;
+
+    newMonth = newMonth + 1;
+    if (newMonth > 12) {
```
Comment (ghost) [inline practice/src/js/controllers/form_input_controller.js]: const TOTAL_MONTHS_OF_YEAR = 12

File: practice/src/js/controllers/page_load_controller.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+  // get date today
+  getDayToDay: function () {
+    const today = new Date();
+    const year = today.getFullYear();
+    const month = String(today.getMonth() + 1).padStart(2, '0');
+    const day = String(today.getDate()).padStart(2, '0');
+    const result = `${day}/${month}/${year}`;
+    return result;
+  }
```
Comment (ghost) [inline practice/src/js/controllers/page_load_controller.js:35]: const date = new Date(Date.UTC(2012, 11, 12, 3, 0, 0));

// toLocaleDateString() without arguments depends on the implementation,
// the default locale, and the default time zone
console.log(date.toLocaleDateString());
// "12/11/2012" if run in en-US locale with time zone America/Los_Angeles

Comment (ghost) [reply practice/src/js/controllers/page_load_controller.js:35]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString

File: practice/src/css/index.css [hunk 11]
```diff
@@ -0,0 +1,10 @@
+@import '/src/css/base/reset.css';
+@import './components/button.css';
+@import './base/variables.css';
+@import './components/title.css';
+@import './base/base.css';
+@import '/src/css/sections/form_input.css';
+@import '/src/css/components/input_slider.css';
+@import '/src/css/sections/table_result.css';
+@import '/src/css/sections/modal.css';
+@import '/src/css/sections/calculate_container.css';
```
Comment (ghost) [inline practice/src/css/index.css]: why css come from many positions?

File: practice/src/index.html [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+  <title>Loan caculator</title>
+</head>
+
+<body>
+  <!-- Header -->
+  <h1 class="title">Loan repayment schedule</h1>
+  <!-- button -->
+  <div class="button-container">
+    <button class="btn active" onclick="activeButton(0)">Decreasing balance sheet</button>
+    <button class="btn" onclick="activeButton(1)">Table of fixed monthly payments</button>
```
Comment (ghost) [inline practice/src/index.html]: activeButton(1)?

File: practice/src/js/views/page_load_view.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+  },
+
+  // open and close modal
+  openCloseModal: function () {
+    const modal = document.querySelector('.modal-container');
+    modal.classList.toggle('modal--visible');
+  },
+
+  //change Loan rate by Loan amount
+  calculateLoanRateByLoanAmount: function () {
```
Comment (ghost) [inline practice/src/js/views/page_load_view.js]: refactor this function

File: practice/src/js/views/form_input_view.js [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+    // get value when user choose date
+    $("#datepicker").on("change", () => {
+
+      getValue();
+    })
+  },
+
+  // set value of table result
+  setValueOfTableResult: function (totaInterestPayple, minMonthlyPayment, maxMonthlyPayment, totalInterest, totalOrigin, result) {
```
Comment (ghost) [inline practice/src/js/views/form_input_view.js]: can you split it to multi function

File: practice/src/js/models/loan_model.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+  disbursementDate
+
+  constructor(propertyValue, loanAmount, loanTerm, interestRate, disbursementDate) {
+    this.propertyValue = propertyValue;
+    this.loanAmount = loanAmount;
+    this.loanTerm = loanTerm;
+    this.interestRate = interestRate;
+    this.disbursementDate = disbursementDate;
+  }
+}
```
Comment (tungtran-zb) [inline practice/src/js/models/loan_model.js:24]: Why class don't have any method?

File: practice/src/js/models/loan_payment_model.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+  repaymentPeriod
+
+  constructor(origin, remainningOriginalAmount, interest, repaymentPeriod) {
+    this.origin = origin;
+    this.remainningOriginalAmount = remainningOriginalAmount;
+    this.interest = interest;
+    this.repaymentPeriod = repaymentPeriod;
+    this.toralPrincipalAndInterest = origin + interest;
+  }
+}
```
Comment (tungtran-zb) [inline practice/src/js/models/loan_payment_model.js:13]: Why class don't have any method?

File: practice/src/js/views/form_input_view.js [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+    const loanTermError = document.querySelector('#loan-term-error');
+    const interestRateError = document.querySelector('#interest-rate-error');
+    const disbursementDateError = document.querySelector('#date-time-error');
+
+    // set message to error label
+    propertyValueError.textContent = errorsAndMessages.errorMessages[0];
+    loanAmountError.textContent = errorsAndMessages.errorMessages[1];
+    loanTermError.textContent = errorsAndMessages.errorMessages[2];
+    interestRateError.textContent = errorsAndMessages.errorMessages[3];
+    disbursementDateError.textContent = errorsAndMessages.errorMessages[4];
```
Comment (tungtran-zb) [inline practice/src/js/views/form_input_view.js]: really confuse me about index? Can you handle it better?

File: practice/src/js/views/form_input_view.js [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+  // function to set value to row of table result
+  setValueToTableRow: function (modalTable, item, number) {
+
+    const row = modalTable.insertRow();
+    const ordinalNumber = row.insertCell(0);
+    const repaymentPeriod = row.insertCell(1);
+    const remainingOriginalAmount = row.insertCell(2);
+    const origin = row.insertCell(3);
+    const interest = row.insertCell(4);
+    const total = row.insertCell(5);
```
Comment (tungtran-zb) [inline practice/src/js/views/form_input_view.js]: you should create const for number form 0 => 5
Example
const ORDINAL_ROW = 0

File: practice/src/js/views/form_input_view.js [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+
+    ordinalNumber.textContent = number;
+    repaymentPeriod.textContent = item.repaymentPeriod;
+    !isNaN(item.remainningOriginalAmount) ? remainingOriginalAmount.textContent = FormInputController.reformater(Math.round(item.remainningOriginalAmount)) : remainingOriginalAmount.textContent = '0';
+    !isNaN(item.origin) ? origin.textContent = FormInputController.reformater(Math.round(item.origin)) : origin.textContent = '0';
+    !isNaN(item.interest) ? interest.textContent = FormInputController.reformater(Math.round(item.interest)) : interest.textContent = '0';
+    !isNaN(item.toralPrincipalAndInterest) ? total.textContent = FormInputController.reformater(Math.round(item.toralPrincipalAndInterest)) : total.textContent = '0';
```
Comment (ghost) [inline practice/src/js/views/form_input_view.js]: Can you handle it simple? for all !isNaN() ?

File: practice/src/js/views/page_load_view.js [hunk 20]
```diff
@@ -0,0 +1,126 @@
+import { PageLoadController } from "../controllers/page_load_controller";
+import flatpickr from 'flatpickr';
+import 'flatpickr/dist/flatpickr.min.css';
```
Comment (ghost) [inline practice/src/js/views/page_load_view.js]: why you impor css here?

File: practice/src/js/views/page_load_view.js [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+      button.classList.toggle('active', i === index);
+    });
+    underLine.style.left = `${index * 50}%`;
+    lineWidth === '50%' ? underLine.style.width = '44%' : underLine.style.width = '50%';
+
+  },
+
+  initializeDatePicker: function () {
+    flatpickr("#datepicker", {
+      dateFormat: "d/m/Y" // Định dạng Ngày/Tháng/Năm
```
Comment (ghost) [inline practice/src/js/views/page_load_view.js]: use English for comments

File: practice/src/index.html [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+
+    </div>
+  </div>
+  <script type="module" src="./js/views/page_load_view.js"></script>
+  <script type="module" src="./js/controllers/page_load_controller.js"></script>
+  <script type="module" src="./js/models/loan_payment_model.js"></script>
+  <script type="module" src="./js/models/loan_model.js"></script>
+  <script type="module" src="./js/views/form_input_view.js"></script>
+  <script type="module" src="./js/controllers/form_input_controller.js"></script>
+  <script type="module" src="./js/main.js"></script>
```
Comment (ghost) [inline practice/src/index.html:171]: Can you make it simple, just import ./js/main.js like css
Too many scripts here

File: practice/src/js/controllers/page_load_controller.js [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+  // reformat number input
+  reformatNumber: function (idElement) {
+    document.addEventListener('DOMContentLoaded', () => {
+      const number_input = document.querySelector(idElement);
+
+      number_input.addEventListener('input', (e) => {
+        let value = e.target.value;
+        value = value.replace(/,/g, '');
+        if (!isNaN(value) && value !== '') {
+          value = Number(value).toLocaleString('en');
```
Comment (ghost) [inline practice/src/js/controllers/page_load_controller.js]: Why number is en
you support viet nam people?

File: practice/src/js/views/page_load_view.js [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+    const INDEX_OF_BUTTON = 0;
+
+    button.addEventListener('click', () => {
+      this.activeButton(INDEX_OF_BUTTON);
+    });
+  },
+
+  // active button left (decreasing balance sheet)
+  activeButtonRight: function () {
+    const button = document.querySelector("#btn-active-1");
```
Comment (ghost) [inline practice/src/js/views/page_load_view.js]: #btn-active-1 => not a good name

File: practice/src/js/controllers/form_input_controller.js [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+        totaInterestPayple = totaInterestPayple + resultRecord.toralPrincipalAndInterest;
+        if (minMonthlyPayment > resultRecord.toralPrincipalAndInterest) {
+          minMonthlyPayment = resultRecord.toralPrincipalAndInterest;
+        };
+        if (maxMonthlyPayment < resultRecord.toralPrincipalAndInterest) {
+          maxMonthlyPayment = resultRecord.toralPrincipalAndInterest;
+        }
+        result.push(resultRecord);
+      }
+    }
```
Comment (ghost) [inline practice/src/js/controllers/form_input_controller.js]: So complicated here, please help refactor

File: practice/src/js/views/page_load_view.js [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+  activeButton: function (index) {
+    const buttons = document.querySelectorAll('.btn');
+    const underLine = document.querySelector('.line-under');
+    const lineWidth = underLine.style.width;
+
+    buttons.forEach((button, i) => {
+      button.classList.toggle('active', i === index);
+    });
+    underLine.style.left = `${index * 50}%`;
+    lineWidth === '50%' ? underLine.style.width = '44%' : underLine.style.width = '50%';
```
Comment (ghost) [inline practice/src/js/views/page_load_view.js]: Should not use js to style, please create class to do it

File: practice/src/js/controllers/form_input_controller.js [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+  isDateInPast: function (valueOfDisbursementDate) {
+    const date = PageLoadController.getDayToDay();
+    const [day, month, year] = date.split('/').map(Number);
+    const [inputDay, inputMonth, inputYear] = valueOfDisbursementDate.split('/').map(Number);
+
+    if (inputYear < year) return true;
+    if (inputYear === year && inputMonth < month) return true;
+    if (inputYear === year && inputMonth === month && inputDay < day) return true;
+    return false;
+  },
```
Comment (ghost) [inline practice/src/js/controllers/form_input_controller.js]: Example

```
isDateInPast: function (valueOfDisbursementDate) {
    const currentDate = PageLoadController.getDayToDay().split('/').map(Number);
    const inputDate = valueOfDisbursementDate.split('/').map(Number);

    const [currentDay, currentMonth, currentYear] = currentDate;
    const [inputDay, inputMonth, inputYear] = inputDate;

    const current = new Date(currentYear, currentMonth - 1, currentDay);
    const input = new Date(inputYear, inputMonth - 1, inputDay);

    return input < current;
},
````

**Review by tungtran-agilityio — APPROVED** [review]

**Pull request merged**